modules/building-memcached-operator-using-osdk.adoc

```diff
@@ -3,441 +3,523 @@
+. Generate and update CRD manifests:
 +
 [source,terminal]
 ----
-$ oc get deployment
+$ make manifests
 ----
 +
-.Example output
-[source,terminal]
-----
-NAME                     DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
-memcached-operator       1         1         1            1           1m
-----
+This Makefile target invokes the `controller-gen` utility to generate the CRD manifests in the `config/crd/bases/cache.example.com_memcacheds.yaml` file.
 
-... _Option B:_ Running locally outside the cluster.
+.. Optional: Add custom validation to your CRD.
 +
-This method is preferred during development cycle to deploy and test faster.
+OpenAPI v3.0 schemas are added to CRD manifests in the `spec.validation` block when the manifests are generated. This validation block allows Kubernetes to validate the properties in a Memcached CR when it is created or updated.
 +
-Run the Operator locally with the default Kubernetes configuration file present at `$HOME/.kube/config`:
+As an Operator author, you can use Kubebuilder markers (annotations) to configure custom validations for your API. These markers must always have a `+kubebuilder:validation` prefix. For example, adding an enum-type specification can be done by adding the following marker:
 +
-[source,terminal]
+[source,go]
 ----
-$ operator-sdk run --local --namespace=default
+// +kubebuilder:validation:Enum=Lion;Wolf;Dragon
+type Alias string
 ----
 +
-You can use a specific `kubeconfig` using the flag `--kubeconfig=<path/to/kubeconfig>`.
-
-. *Verify that the Operator can deploy a Memcached application* by creating a `Memcached` CR.
-
-.. Create the example `Memcached` CR that was generated at `deploy/crds/cache_v1alpha1_memcached_cr.yaml`.
-
-.. View the file:
+Usage of markers in API code is discussed in the Kubebuilder link:https://book.kubebuilder.io/reference/generating-crd.html[Generating CRDs] and link:https://book.kubebuilder.io/reference/markers.html[Markers for Config/Code Generation] documentation. A full list of OpenAPIv3 validation markers is also available in the Kubebuilder link:https://book.kubebuilder.io/reference/markers/crd-validation.html[CRD Validation] documentation.
 +
-[source,terminal]
-----
-$ cat deploy/crds/cache_v1alpha1_memcached_cr.yaml
-----
+If you add any custom validations, run the following command to update the OpenAPI validation section for the CRD:
 +
-.Example output
 [source,terminal]
 ----
-apiVersion: "cache.example.com/v1alpha1"
-kind: "Memcached"
-metadata:
-  name: "example-memcached"
-spec:
-  size: 3
+$ make manifests
 ----
 
-.. Create the object:
+. After creating a new API and controller, you can implement the controller logic. For this example, replace the generated controller file `controllers/memcached_controller.go` with following example implementation:
 +
-[source,terminal]
-----
-$ oc apply -f deploy/crds/cache_v1alpha1_memcached_cr.yaml
+.Example `memcached_controller.go`
```

Comment:
Same code block in box behavior here. I wonder if it's this preview build--is it working locally?